util/personalized.py

```diff
@@ -0,0 +1,252 @@
+    to the assigned exercise instance and directory "personal".
+    '''
+    return os.path.join(settings.PERSONALIZED_CONTENT_PATH,
+        course["key"], "users", userid, exercise["key"])
+
+
+def pregenerated_exercises_directory_path(course, exercise):
+    '''
+    Return path to the directory of pregenerated exercise instances
+    (instances are directories under this directory).
+    '''
+    return os.path.join(settings.PERSONALIZED_CONTENT_PATH,
+        course["key"], "pregenerated", exercise["key"])
+
+
+def prepare_user_personal_directory(course, exercise, userid):
+    '''
+    Create the personal directory of the user (unless the directory already exists).
+    '''
+    user_dir = user_personal_directory_path(course, exercise, userid)
+    personal_dir = os.path.join(user_dir, 'personal')
+    # Create empty directory unless it already exists
+    if not os.path.exists(personal_dir):
+        try:
+            os.makedirs(personal_dir)
+        except OSError:
+            pass # someone else created the dir after executing the if condition
+
+
+def prepare_pregenerated_exercises_directory(course, exercise):
+    '''
+    Create the base directory for pregenerated exercise instances.
+    '''
+    pregen_dir = pregenerated_exercises_directory_path(course, exercise)
+    # Create empty directory unless it already exists
+    try:
+        os.makedirs(pregen_dir)
+    except OSError:
+        return # it exists
+
+
+def pregenerated_exercise_instances(course, exercise):
+    '''
+    Return a list of the existing pregenerated exercises instances (directory names).
+    '''
+    pregenerated_dir = pregenerated_exercises_directory_path(course, exercise)
+    try:
+        # return a list of directory names (not full paths)
+        return [instance_dir for instance_dir in os.listdir(pregenerated_dir) 
+                if os.path.isdir(os.path.join(pregenerated_dir, instance_dir))]
+    except OSError:
+        return [] # pregenerated_dir does not exist
+
+
+def select_generated_exercise_instance(course, exercise, userids_str, submission_number):
+    '''
+    Return path to the generated exercise instance that is assigned to the user(s).
+    The instance may change as more submissions are made if the exercise has
+    enabled regeneration.
+    '''
+    pregenerated_dir = pregenerated_exercises_directory_path(course, exercise)
+    userids = sum(user_ids_from_string(userids_str))
+    instances = pregenerated_exercise_instances(course, exercise)
+    num_instances = len(instances)
+    if num_instances == 0:
+        raise access.config.ConfigError("Exercise is personalized but no exercise instances have been pregenerated")
+    
+    if "max_submissions_before_regeneration" in exercise:
+        # the generated exercise may be regenerated after submitting certain amount of times
+        instance = (userids + ((submission_number - 1) // exercise["max_submissions_before_regeneration"]) *
```

Comment:
Yes, there are obvious problems with this. Possible solution:
instances = list(range(num_instances))
r = random.Random(userids)
r.shuffle(instances)
instance = instances[((submission_number - 1) // exercise["max_submissions_before_regeneration"]) % num_instances]